Assignment2 - Make Memorize Game

kuk6933 opened the pull request:
Week2 Assignment
MVVM 구조에 유의하여 Memorize Game 구현하기
Model - (Theme, MemorizeGame)
View -(MemorizeGameView 등 View 폴더 내 파일들)
ViewModel- MemorizeGameDealer
이렇게 나누어 구현했습니다.
Week1 Assignment때 피드백 받은대로 coding convention을 준수하는것을 가장 신경썼습니다.

부족한 점
-> MVVM구조를 준수하려고 노력했는데 마지막까지도 어떤 코드를 어느 곳에 넣을지 확신을 가지지 못했습니다. MVVM 구조에 대한 이해가 더 필요함을 느꼈습니다.

-> 5가지 접근제어에 대한 이론은 알고있었어도 적용에 대한 어려움이 많았습니다. 

-> 고차함수나 클로저 등 Swift 문법에 더 익숙해질 필요가 있다고 느꼈습니다.
기능은 모두 구현했습니다! 

-> 네이밍에 좀 더 신경 써야 할 것 같습니다.

-> Model/MemorizeGame의 IsTwoCardsFaceUp() 메서드가 좀 지저분해 보 
     이는데 개선방법을 떠올리지 못했습니다

kuk6933 committed:
Assignment2 Setting

kuk6933 committed:
Make some change
-Delete button file
-Add theme
-Make display
-Make ThemeNameView
-Make NewStartButton

kuk6933 committed:
temporary commint

kuk6933 committed:
newGame button function

kuk6933 committed:
Add score function
Update for coding convention

kuk6933 committed:
edit CardColor file
-Model must not import SwiftUI

kuk6933 committed:
Function implementation complete

kuk6933 committed:
Prevent choosing same card

kuk6933 committed:
Function Update
-Even if user clicks New Game Button, if theme is same, user get same cards -> Edit changeTheme Function

kuk6933 committed:
bug fix
-카드리스트에 한 카드가 한쌍이상 나오는 case가 있어 수정했습니다

kuk6933 committed:
Update ViewModel, Model file name

kuk6933 committed:
Delete useless code

kuk6933 committed:
edit

kuk6933 committed:
Update SuccessView UI

kuk6933 committed:
Recovery

kuk6933 committed:
please

kuk6933 committed:
temp

kuk6933 committed:
Read me

gen-com commented on week2/Memorize/Memorize/MemorizeApp.swift:
```diff
@@ -0,0 +1,13 @@
+
+@main
+struct MemorizeApp: App {
+    var memorizeGame = MemorizeGameDealer()
```
접근제어 !

kuk6933 replied:
private로 수정했습니다

gen-com commented on week2/Memorize/Memorize/ContentView.swift:
```diff
@@ -0,0 +1,36 @@
+import SwiftUI
+
+struct ContentView: View {
```
ContentView보다 더 좋은 이름이 있지 않을까요 ?

kuk6933 replied:
수정했습니다!

gen-com commented on week2/Memorize/Memorize/Thema.swift:
```diff
@@ -0,0 +1,8 @@
+//  Created by ohhyeongseok on 2022/07/26.
+//
+
+import Foundation
```
불필요한 파일 삭제해주세요.

kuk6933 replied:
넵

gen-com commented on week2/Memorize/Memorize/Theme.swift:
```diff
@@ -0,0 +1,20 @@
+
+import SwiftUI
+
+struct Theme: View {
```
여기도 불필요한 파일이라면 지워주세요.

kuk6933 replied:
넵

gen-com commented on week2/Memorize/Memorize/ThemeNameView.swift:
```diff
@@ -0,0 +1,18 @@
+    @EnvironmentObject var memorizeGame: MemorizeGameDealer
+    
+    var body: some View {
+        Text("\(memorizeGame.currentTheme.rawValue)")
```
memorizeGame.currentTheme.rawValue의 값은 String입니다. 문자열 보간법이 필요했을까요 ?
ViewModel이 해석 역할을 더 잘 할 수 있지 않을까요 ? (memorizeGame.currentThemeName)

kuk6933 replied:
그렇네요,, String인데 문자열 보간법이라니
ViewModel에 currentThemeName 프로퍼티를 추가해서 말씀하신 방법으로 수정했습니다!

gen-com commented on week2/Memorize/Memorize/ContentView.swift:
```diff
@@ -0,0 +1,36 @@
+    @EnvironmentObject var memorizeGame: MemorizeGameDealer
+    
+    var body: some View {
+        if !memorizeGame.checkGameIsDone() {
```
상위 뷰에서 foregroundColor 값을 줬다면 하위 뷰에 기본적으로 다 적용되지 않을까요 ?

kuk6933 replied:
상위뷰에서 foregroundColor 값을 주도록 수정했습니다

gen-com commented on week2/Memorize/Memorize/CardListView.swift:
```diff
@@ -0,0 +1,17 @@
+    }
+}
+
+
```
아래에도 의미없는 공백에 신경써주세요 !

kuk6933 replied:
넵 아래도 주의하겠습니다

gen-com commented on week2/Memorize/Memorize/CardListView.swift:
```diff
@@ -0,0 +1,17 @@
+    var body: some View {
+        ScrollView {
+            LazyVGrid(columns: [GridItem(.adaptive(minimum: 65))]) {
+                ForEach(memorizeGame.cards, id: \.self.id) { card in
```
`\.self`는 무엇일까요 ?

kuk6933 replied:
의미없는 중복이기에 하나 삭제했습니다!

gen-com commented on week2/Memorize/Memorize/NewGameButton.swift:
```diff
@@ -0,0 +1,17 @@
+    
+    var body: some View {
+        Button("New Game") {
+            memorizeGame.changeTheme(nextTheme: Theme.randomTheme)
```
새로운 게임을 원한다 -> 테마를 바꾸자 ! 보다는
새로운 게임을 원한다 -> 새로 게임을 시작하자(테마는 자동 변경) 가 더 자연스럽지 않은가요 ?
굳이 매개변수를 전달해야하는 이유도 모르겠습니다.

kuk6933 replied:
네이밍에 더 신경써야겠습니다..
말씀처럼 굳이 View에서 매개변수로 넘겨주지않고 viewModel에서 변수를 만들어 사용하는 방식으로 수정했습니다

gen-com commented on week2/Memorize/Memorize/Model/CardColor.swift:
```diff
@@ -0,0 +1,9 @@
+enum CardColor {
```
네임스페이스가 아깝습니다. Nested 형태를 고려해볼까요 ?

kuk6933 replied:
Theme에 따라 결정되므로 Theme에 Nested하도록 수정했습니다! 
파일을 분리하면 깔끔해보여 좋을것이라 생각했는데 Nested가 가능한 경우라면 하는게 좋을까요?

rose6649 replied:
저는 CardColor가 혼자 나와있을 이유가 있다면 따로 분리하는것이 좋을 것 같고,
Card에 종속되어있는 거라면 Card안에 Color를 표기하는게 더 좋지 않을까 생각되네요!

(말하신대로, Theme따라 결정되는 부분이니까!)

kuk6933 replied:
곰곰이 생각해보니 Theme에 따라 결정되고 ThemeNameView, ScoreView에도 적용되므로 ThemeColor가 더 적합할것 같아 CardColor를  ThemeColor로 수정하고 Theme에 Nested되도록 수정했습니다!

gen-com commented on week2/Memorize/Memorize/Model/Theme.swift:
```diff
@@ -0,0 +1,62 @@
+enum Theme: String, CaseIterable {
+    case animal = "animal"
```
case 명과 rawValue(String) 값이 같다면` = "something"`은 하지 않아도 됩니다.

kuk6933 replied:
감사합니다! 수정했습니다

gen-com commented on week2/Memorize/Memorize/Model/MemorizeGame.swift:
```diff
@@ -0,0 +1,79 @@
+        cards = cards.shuffled()
+    }
+    
+    class Card: Identifiable, ObservableObject {
```
각 카드는 왜 ViewModel 처럼 관리가 되나요 ?
MemorizeGameDealer 가 ViewModel이고 각 카드 원본에 대한 관리를 할 수 있지 않나요 ?

kuk6933 replied:
Class -> Struct로 수정하고
MemorizeGameDealer를 이용하여 원본을 관리하도록 수정하였습니다

gen-com commented on week2/Memorize/Memorize/Model/MemorizeGame.swift:
```diff
@@ -0,0 +1,79 @@
+            }
+        }
+    }
+    mutating func changeTheme(numberOfPairsOfCards: Int, nextTheme: Theme, createContent: (Int) -> CardContent) {
```
함수는 공백으로 구분하는 것을 권장합니다.
init 코드와 중복되지 않나요 ? 중복을 해소할 수는 없을까요 ?

kuk6933 replied:
공통 코드를 함수로 만들어 중복을 해소했습니디

gen-com commented on week2/Memorize/Memorize/MemorizeGameDealer.swift:
```diff
@@ -0,0 +1,65 @@
+    var themeColor: Color {
+        switch currentTheme.cardColor {
+        case .red:
+            return Color.red
```
`var themeColor: Color` 여기서 반환 타입을 유추할 수 있지 않을까요 ? Color는 생략 가능 합니다. 모든 케이스 마찬가지

kuk6933 replied:
수정했습니다!

gen-com commented on week2/Memorize/Memorize/CardView.swift:
```diff
@@ -0,0 +1,35 @@
+            if !card.isMatched {
+                memorizeGame.choose(card: card)
+            }
+            
```
공백

kuk6933 replied:
넵 수정했습니다

gen-com commented on week2/Memorize/Memorize/CardView.swift:
```diff
@@ -0,0 +1,35 @@
+
+struct CardView: View {
+    @EnvironmentObject var memorizeGame: MemorizeGameDealer
+    @ObservedObject var card: MemorizeGame<String>.Card
```
memorizeGame으로부터 카드의 정보를 읽어올 수 있지 않나요 ?

kuk6933 replied:
넵 Card를 Struct로 고치면서 수정했습니다!

gen-com commented on week2/Memorize/Memorize/CardView.swift:
```diff
@@ -0,0 +1,35 @@
+        CardView(card: MemorizeGame<String>.Card(content: "aa", id: 1))
+    }
+}
+
```
공백

kuk6933 replied:
삭제했습니다

gen-com commented on week2/Memorize/Memorize/MemorizeGameDealer.swift:
```diff
@@ -0,0 +1,65 @@
+import SwiftUI
+
+class MemorizeGameDealer: ObservableObject {
+    @Published private var model = createEmojiMemoryGame(currentTheme: Theme.randomTheme)
```
model보다 적합한 이름이 있지 않을까요 ?

kuk6933 replied:
memorizeGame의 데이터, 로직을 관리하므로 memorizeGameManager로 수정했습니다

gen-com commented on week2/Memorize/Memorize/Model/MemorizeGame.swift:
```diff
@@ -0,0 +1,79 @@
+
+struct MemorizeGame<CardContent: Equatable> {
+    private(set) var cards: [Card]
+    private(set) var theme: Theme
```
"메모리게임과 다른 모델로 관리해주세요." 라는 과제 조건에 부합하지 않습니다.

kuk6933 replied:
과제 조건에 맞도록 수정했습니다

gen-com commented on week2/Memorize/Memorize/Model/MemorizeGame.swift:
```diff
@@ -0,0 +1,79 @@
+        }
+    }
+    mutating func changeTheme(numberOfPairsOfCards: Int, nextTheme: Theme, createContent: (Int) -> CardContent) {
+        cards = [Card]()
```
init 코드와 중복되지 않나요 ? 중복을 해소할 수는 없을까요 ?

gen-com commented on week2/Memorize/Memorize/Model/MemorizeGame.swift:
```diff
@@ -0,0 +1,79 @@
+struct MemorizeGame<CardContent: Equatable> {
+    private(set) var cards: [Card]
+    private(set) var theme: Theme
+    private(set) var currentOpenedCards: [Card] = []
```
Opened 보다는 우리는 이미 FaceUp이라는 단어를 사용했었죠 ? 통일되면 더 좋지 않을까요 ?

kuk6933 replied:
통일되도록 수정했습니다!

gen-com commented on week2/Memorize/Memorize/Model/MemorizeGame.swift:
```diff
@@ -0,0 +1,79 @@
+    
+    mutating func choose(card: Card) {
+        if let chosenIndex = cards.firstIndex(where: { $0.id == card.id }) {
+            if currentOpenedCards.count == 1 && currentOpenedCards[0].id == card.id {
```
`[0]`와 같은 직접적인 접근은 굉장히 위험합니다. 지양해주세요. 아래 모두 마찬가지 입니다.
로직을 더 개선해 볼 수 있을 것 같습니다.

Hint, 어떤 카드를 누를 때 이미 앞면을 보고있는 카드는 있을 수도 없을 수도 있습니다. 해당 값을 처리하는 좋은 방법이 있지 않나요 ?

kuk6933 replied:
[0]으로 접근하지않고 array.first 로 접근하도록 수정했습니다만 코드가 조금 길어져버렸네요 더 짧게 할 수 있는 방법이 없는지 더 생각해보겠습니다

gen-com commented on week2/Memorize/Memorize/MemorizeGameDealer.swift:
```diff
@@ -0,0 +1,65 @@
+        }
+    }
+    
+    static func createEmojiMemoryGame(currentTheme: Theme) -> MemorizeGame<String> {
```
접근제어 ?

kuk6933 replied:
private로 수정했습니다!

gen-com commented on week2/Memorize/Memorize/MemorizeGameDealer.swift:
```diff
@@ -0,0 +1,65 @@
+    func choose(card: MemorizeGame<String>.Card) {
+        model.choose(card: card)
+    }
+    func checkGameIsDone() -> Bool {
```
호출하는 입장에서 game.isFinished 가 더 잘 읽히지 않나요 ? 그러고 보면 computed 변수로 만들 수도 있겠네요 ?

`model.numberOfMatchedPairsOfCards == model.cards.count` 이 값 자체가 Bool 아닌가요 ? 그대로 반환해도 될 거 같은데요 ?

kuk6933 replied:
아하 파라미터를 받지않고 Bool 타입 한개를 반환하니 computed 변수로 만드는것이 좋겠군요 수정했습니다 
그대로 반환하도록 수정했습니다!

gen-com commented on week2/Memorize/Memorize/Model/Theme.swift:
```diff
@@ -0,0 +1,62 @@
+    static var randomTheme: Theme {
+        return Theme.allCases.randomElement() ?? .emotion
+    }
+    var emojis: [String] {
```
`.shuffled()`가 반복되는 것을 개선할 수 있지 않을까요 ?

kuk6933 replied:
개선하여 수정했습니다!

gen-com reviewed (commented):
1차 피드백 남겼습니다. 빠진 부분도 있을테니, 그 부분도 신경써서 개선했으면 좋겠습니다.

1주차 과제에 비해 신경 쓴 모습이 보여 좋습니다. 좀 더 디테일해지면 더 좋겠죠 ?

kuk6933 committed:
공백 제거/ 불필요한 파일 삭제

rose6649 commented:
커밋메세지랑 커밋단위 조금만 신경써주세요! ;)
please temp read me recovery 
뭔지 궁금해서 들어가봤는데 한줄만 고친것도 있고 뭐 그런것같네요ㅎㅎ

rose6649 commented on week2/Memorize/Memorize/MemorizeGameDealer.swift:
```diff
@@ -0,0 +1,65 @@
+    }
+    
+    static func createEmojiMemoryGame(currentTheme: Theme) -> MemorizeGame<String> {
+        var numberOfPairsOfCards = currentTheme.numberOfPairsOfCardsToShow
+        let constantEmojis = currentTheme.emojis
+        
+        if currentTheme.numberOfPairsOfCardsToShow > currentTheme.emojis.count {
+            numberOfPairsOfCards = currentTheme.emojis.count
+        }
```
네이밍이 길지만, 뭐하는지 이해할 수 있어 좋아요! 앞으로도 좋은 네이밍 기대할게요

kuk6933 replied:
감사합니다! 
아 please temp 이 커밋들은 깃 push할떄 reject되어서 원인을 찾으려고 했던 커밋들이었습니다 ㅠ 커밋 메세지 신경쓰겠습니다!

rose6649 commented on week2/Memorize/Memorize/MemorizeGameDealer.swift:
```diff
@@ -0,0 +1,65 @@
+    
+    static func createEmojiMemoryGame(currentTheme: Theme) -> MemorizeGame<String> {
+        var numberOfPairsOfCards = currentTheme.numberOfPairsOfCardsToShow
+        let constantEmojis = currentTheme.emojis
```
let이 상수인것을 나타내는데, 상수이모지는 쪼끔 어색하지않나? 싶어요 ㅎㅎ

kuk6933 replied:
의미가 중복되네요..  수정했습니다!

rose6649 commented on week2/Memorize/Memorize/MemorizeGameDealer.swift:
```diff
@@ -0,0 +1,65 @@
+    }
+    func changeTheme(nextTheme: Theme) {
+        var numberOfPairsOfCards = nextTheme.numberOfPairsOfCardsToShow
+        let constantEmojis = nextTheme.emojis
+        
+        if nextTheme.numberOfPairsOfCardsToShow > nextTheme.emojis.count {
```
여기서 constantEmojis을 선언하고 nextTheme.emojis.count는 constantEmojis.count로 안하네요.
이유가 있을까요??

kuk6933 replied:
nextThmeme.emojis.count와 constantEmojis.count가 같은 값을 가지기에 로직상 이슈가 없어 신경쓰지 못했던것 같습니다
통일되도록 수정했습니다 !

rose6649 commented on week2/Memorize/Memorize/Model/MemorizeGame.swift:
```diff
@@ -0,0 +1,79 @@
+    private(set) var numberOfMatchedPairsOfCards = 0
+    
+    init(numberOfPairsOfCards: Int, theme: Theme, createContent: (Int) -> CardContent) {
+        cards = [Card]()
```
cards: [Card]를 init()문에서 초기화 하고, currentOpenedCards: [Card] = []는 직접적으로 초기화해주셨네요!
이 둘의 차이를 둔 이유를 알 수 있을까요??

kuk6933 replied:
cards는 예시코드 따라 작성한 코드이고 currentOpendedCards는 제가 추가로 넣은 코드여서 차이가 발생한것 같습니다
통일되도록 수정했습니다!

kuk6933 committed:
리뷰반영-version1
- add Access Control
- unify naming
- edit naming
- Add Group View, ViewModel

kuk6933 committed:
리뷰반영- version2
메모리게임과 Theme를 다른 모델로 관리하도록 수정
MemoryGame 파일에서 init()과 makeCards 함수의 중복되는 부분 빼서 함수로 만들어 중복제거

kuk6933 committed:
Bug fix
게임을 시작할 때 Theme와 card content가 일치하지 않는 경우가 발생해 이를 수정함

kuk6933 committed:
MemorizeGame.choose함수 분리
함수가 너무 길어져 보기가 힘들어져 코드를 분리해서 함수를 만들고 이를 호출하도록 했습니다

kuk6933 committed:
Keep coding convention
- 조건문 공백 수정
- 함수 호출 순서에 따른 순서 조정

kuk6933 committed:
Edit Enum
CardColor -> ThemeColor

kuk6933 committed:
edit name
memorizeGame -> memorizeGameDealer

kuk6933 committed:
공백 삭제

gen-com commented on week2/Memorize/Memorize/Model/MemorizeGame.swift:
```diff
@@ -0,0 +1,95 @@
+import Foundation
+
+struct MemorizeGame<CardContent: Equatable> {
+    private(set) var cards: [Card] = []
```
cards 초기화는 init 메소드에서 수행하므로 디폴트 값을 굳이 주지 않아도 될 거 같네요.

kuk6933 replied:
디폴트 값을 주지 않을 경우 self' used before all stored properties are initialized 메세지가 떠서 디폴트 값을 설정해놨습니다!

gen-com commented on week2/Memorize/Memorize/Model/MemorizeGame.swift:
```diff
@@ -0,0 +1,95 @@
+        var isMatched = false
+        var isFaceUpAtLeastOnce = false
+        let content: CardContent
+        var id: Int
```
id는 변하는 값인가요 ?

kuk6933 replied:
let으로 수정했습니다

gen-com commented on week2/Memorize/Memorize/Model/MemorizeGame.swift:
```diff
@@ -0,0 +1,95 @@
+    private(set) var cards: [Card] = []
+    private(set) var score = 0
+    private var currentOpenedCards: [Card] = []
+    var numberOfMatchedPairsOfCards: Int {
```
해당 값은 외부에서도 사용해야하나요 ?
또 게임 종료는 로직의 일부 아닌가요 ?

kuk6933 replied:
수정했습니다!

gen-com commented on week2/Memorize/Memorize/Model/MemorizeGame.swift:
```diff
@@ -0,0 +1,95 @@
+    }
+}
+
+extension MemorizeGame {
```
여기는 extension으로 구분한 이유는 무엇인가요 ?
그리고 아래 메소드들은 외부에서 호출되지 않는다면 접근제어가 필요하지 않을까요 ?
배열에 대한 직접적인 접근을 하는 것으로 보아 아직 좋은 로직이 떠오르지 않은 모양입니다.

kuk6933 replied:
choose 함수 코드가 길어서 보기편하도록 extension으로 따로 빼놨는데 생각해보니 그러면 다른 함수도 분리해야 통일성을 해치지 않겠네요 수정했습니다

gen-com commented on week2/Memorize/Memorize/Model/MemorizeGame.swift:
```diff
@@ -0,0 +1,95 @@
+        }
+    }
+    
+    mutating func FaceDownTwoCards() {
```
함수는 소문자로 시작합니다.

kuk6933 replied:
넵

gen-com commented on week2/Memorize/Memorize/Model/MemorizeGame.swift:
```diff
@@ -0,0 +1,95 @@
+        }
+    }
+    
+    mutating func IsTwoCardsFaceUp() {
```
소문자로 시작, 접근 제어 ?

kuk6933 replied:
수정했습니다

gen-com commented on week2/Memorize/Memorize/View/CardListView.swift:
```diff
@@ -0,0 +1,15 @@
+    var body: some View {
+        ScrollView {
+            LazyVGrid(columns: [GridItem(.adaptive(minimum: 65))]) {
+                ForEach(memorizeGameDealer.cards, id: \.id) { card in
```
id 부분 생략 가능합니다. 참고해요.

kuk6933 replied:
넵

gen-com commented on week2/Memorize/Memorize/View/CardView.swift:
```diff
@@ -0,0 +1,33 @@
+            if !card.isMatched {
+                memorizeGameDealer.choose(card: card)
+            }
+            
```
공백

kuk6933 replied:
넵

gen-com commented on week2/Memorize/Memorize/View/CardView.swift:
```diff
@@ -0,0 +1,33 @@
+
+struct CardView: View {
+    @EnvironmentObject var memorizeGameDealer: MemorizeGameDealer
+    var card: MemorizeGame<String>.Card
```
뷰에서 카드가 변경될 수 있어야 하는  값인가요 ?

kuk6933 replied:
let으로 수정했습니다

gen-com commented on week2/Memorize/Memorize/ViewModel/MemorizeGameDealer.swift:
```diff
@@ -0,0 +1,68 @@
+
+class MemorizeGameDealer: ObservableObject {
+    @Published private var memorizeGameManager = createEmojiMemoryGame()
+    static private var theme = Theme.randomTheme
```
`private static` 으로 통일해주세요.

kuk6933 replied:
넵

gen-com commented on week2/Memorize/Memorize/ViewModel/MemorizeGameDealer.swift:
```diff
@@ -0,0 +1,68 @@
+    }
+    
+    private static func createEmojiMemoryGame() -> MemorizeGame<String> {
+        Self.theme = .randomTheme
```
Self는 무엇일까요 ? Self는 필요할까요 ?

kuk6933 replied:
Self는 타입의 이름을 반복해서 쓰는 대신에 현재 타입을 편하게 가리킬수 있도록 해준다고 해서 사용했습니다.

rose6649 replied:
self와 Self의 차이점은 뭘까요?
그렇다면 theme을 static으로 둔 이유는 무엇일까요?

kuk6933 replied:
self는 instance에서 자기 자신을 가리키고
Self는 타입을 가리킵니다
theme를 static으로 둔 이유는 creaateEmojiMemoryGame을 static으로 만들었는데 이에 넣어주기 위해 만들었었는데 생각해보니 그러면 createEmojiMemoryGame에서 Self.를 안써줘도 되겠군요 수정했습니다 !

gen-com reviewed (commented):
조금 더 개선이 필요한 것 같습니다. 치밀한 설계와 꼼꼼함을 잊지 마세요.

kuk6933 committed:
Reflect review-2
- access control
- coding convention
- edit struct

rose6649 commented on week2/Memorize/Memorize/Model/MemorizeGame.swift:
```diff
@@ -0,0 +1,92 @@
+        isTwoCardsFaceUp()
+    }
+    
+    mutating private func faceUpCardAftercheck(card: Card) {
```
AfterCheck

kuk6933 replied:
넵

rose6649 commented on week2/Memorize/Memorize/Model/MemorizeGame.swift:
```diff
@@ -0,0 +1,92 @@
+    mutating func choose(card: Card) {
+        faceUpCardAftercheck(card: card)
+        faceDownTwoCards()
+        isTwoCardsFaceUp()
```
보통 is~ can ~ 이런 형의 네이밍으로 시작되는 경우에는 Bool값을 가지는 경우가 많아요
이건 함수형이니, 이 함수가 전체적으로 어떤 기능을 하는지에 더 초점을 맞춰서 네이밍을 해보면 좋을 것 같아요

leejun6694 replied:
method 가 어떤 동작을 하는지, 동사로 네이밍을 시작해볼까요?

kuk6933 replied:
넵!

rose6649 commented on week2/Memorize/Memorize/Model/Theme.swift:
```diff
@@ -0,0 +1,75 @@
+        }
+        return emoji.shuffled()
+    }
+    var themeColor: ThemeColor {
```
Theme안에서 themeColor로 중복네이밍

kuk6933 replied:
넵

rose6649 commented on week2/Memorize/Memorize/Model/Theme.swift:
```diff
@@ -0,0 +1,75 @@
+    }
+}
+
+extension Theme {
+    enum ThemeColor {
+        case red
+        case orange
+        case yellow
+        case green
+        case blue
+        case indigo
+        case purple
+    }
+}
```
ThemeColor만 왜 extension으로 빼온건지 이유를 들을 수 있을까요??

kuk6933 replied:
좀더 보기 편하지 않을까 해서 분리해보았습니다

leejun6694 commented on week2/Memorize/Memorize/Model/MemorizeGame.swift:
```diff
@@ -0,0 +1,92 @@
+    }
+    
+    mutating private func isTwoCardsFaceUp() {
+        if currentOpenedCards.count == 2 {
```
else 가 없는 경우 guard 로 작성하면 가독성을 증가시킬 수 있습니다 : )

kuk6933 replied:
넵

leejun6694 commented on week2/Memorize/Memorize/Model/MemorizeGame.swift:
```diff
@@ -0,0 +1,92 @@
+                if let chosenIndex = cards.firstIndex(where: { $0.id == currentOpenedCards.first?.id }) {
+                    cards[chosenIndex].isFaceUpAtLeastOnce = true
+                }
+                if let chosenIndex = cards.firstIndex(where: { $0.id == currentOpenedCards.reversed().first?.id }) {
```
if 문들이 이어지고 있는데, else if 로 처리는 안되는지 검토 부탁해요~

kuk6933 replied:
넵

leejun6694 commented on week2/Memorize/Memorize/ViewModel/MemorizeGameDealer.swift:
```diff
@@ -0,0 +1,68 @@
+        return memorizeGameManager.isFinished
+    }
+    var themeColor: Color {
+        switch currentTheme.themeColor {
```
`return currentTheme.themeColor`

kuk6933 replied:
굳이 enum ThemeColor를 만들어서 ThemeColor로 return하는것보다 Color를 return하게하고 이렇게 수정하는게 더 낫겠네요 수정했습니다!

kuk6933 committed:
Reflect review-3
if문 중복 삭제
naming 수정

kuk6933 committed:
recovery
-Model과 UI분리

kuk6933 committed:
Assignment3- Make Calculator
- Setting
- implement functions
- set fundamental LayOut

kuk6933 committed:
Update Button UI layout
Edit ButtonSize Using the UIScreen class

kuk6933 committed:
Add divide by zero exception

kuk6933 committed:
옵셔널 강제추출 삭제
logic 수정

kuk6933 committed:
Apply minimum scale factor

kuk6933 committed:
keep coding convention

kuk6933 committed:
recovery

kuk6933 committed:
View 전체에서 gesture 기능

kuk6933 committed:
move Button to viewModel
edit backWhenDragged function

kuk6933 committed:
Edit buttonStyle

kuk6933 committed:
Add MARK

kuk6933 committed:
extend gesture range

kuk6933 committed:
Week5
Edit insertComma

kuk6933 committed:
Make input lower than 1,000,000,000

kuk6933 committed:
Saparate Extensions

kuk6933 committed:
Edit naming
Keep coding convention
Delete unnecessary code

kuk6933 committed:
Keep coding convention

kuk6933 committed:
Week5 -> Week4

gen-com commented on week4/Calculator/Calculator/Extensions/Formatter+Extension.swift:
```diff
@@ -0,0 +1,18 @@
+import Foundation
+
+extension Formatter {
+    static let exponentialNotation: NumberFormatter = {
```
작아지는 방향은 없나요 ?

gen-com commented on week4/Calculator/Calculator/Extensions/Numeric+Extension.swift:
```diff
@@ -0,0 +1,14 @@
+
+extension Numeric {
+    var exponentialNotation: String {
+        return Formatter.exponentialNotation.string(for: self) ?? ""
```
앞의 값이 nil일 때 공백 문자열을 보내도 아무 상관 없는 건가요 ?

gen-com commented on week4/Calculator/Calculator/Extensions/String+Extension.swift:
```diff
@@ -0,0 +1,28 @@
+import Foundation
+
+extension String {
+    var insertComma: String {
```
``` swift
var decimalFormat: String {
    let numberFormatter = NumberFormatter()
    numberFormatter.numberStyle = .decimal
    let decimalFractionComponents = components(separatedBy: ".")
    guard let decimalString = decimalFractionComponents.first,
          let decimal = Int(decimalString),
          var decimalformat = numberFormatter.string(from: NSNumber(value: decimal))
    else { return "오류" }
    if 1 < decimalFractionComponents.count, let fractionString = decimalFractionComponents.last {
        decimalformat += ".\(fractionString)"
    }
    return decimalformat
}
```

이렇게는 어떤가요 ?

gen-com commented on week4/Calculator/Calculator/Style/ButtonStyle.swift:
```diff
@@ -0,0 +1,42 @@
+    let isZero: Bool
+
+    func makeBody(configuration: Self.Configuration) -> some View {
+        if isZero {
```
``` swift
configuration.label
    .frame(
        maxWidth: UIScreen.main.bounds.size.width / (isZero ? 2.0 : 6.5),
        maxHeight: UIScreen.main.bounds.size.height / 12,
        alignment: isZero ? .leading : .center
    )
    .font(.title)
    .padding()
    .background(buttonColor.background)
    .foregroundColor(buttonColor.foreground)
    .overlay {
        if configuration.isPressed {
            Color(white: 1.0, opacity: 0.3)
        }
    }
    .clipShape(Capsule())
```

leejun6694 replied:
중복 코드를 줄여봐요~

gen-com commented on week4/Calculator/Calculator/Style/ButtonStyle.swift:
```diff
@@ -0,0 +1,42 @@
+//
+import SwiftUI
+
+struct CalculateButtonStyle: ButtonStyle {
```
ButtonStyle 활용하는 것은 좋네요 👍🏻

gen-com commented on week4/Calculator/Calculator/View/CalculatorView.swift:
```diff
@@ -0,0 +1,32 @@
+            PadView()
+        }
+        .contentShape(Rectangle())
+        .gesture(DragGesture(minimumDistance: 10, coordinateSpace: .local)
```
이렇게 하면 VStack 전체에 대해 드래그 제스처를 인지하게 됩니다.

gen-com commented on week4/Calculator/Calculator/View/CalculatorView.swift:
```diff
@@ -0,0 +1,32 @@
+    var body: some View {
+        VStack {
+            Spacer()
+            ScreenView()
```
``` swift
private func swipeRight() -> some Gesture {
    DragGesture(minimumDistance: 10, coordinateSpace: .local)
        .onEnded { drag in
            if 0 < drag.translation.width {
                calculatorManager.undoWhenSwiped()
            }
        }
}
```
위와 같이 제스처를 메서드로 처리할 수 있습니다.
``` swift
ScreenView()
    .contentShape(Rectangle())
    .gesture(swipeRight())
```
이렇게 쓸 수 있겠죠 ?
contentShape은 왜 필요하고, translation은 무엇일까요 ?

gen-com commented on week4/Calculator/Calculator/View/EachButton.swift:
```diff
@@ -0,0 +1,27 @@
+        ZStack {
+            Button(button.appearance) {
+                calculatorManager.touchButton(button)
+            } .buttonStyle(CalculateButtonStyle(buttonColor: calculatorManager.buttonColor(button), isZero: button.appearance == "0"))
```
``` swift
.buttonStyle(
    CalculateButtonStyle(
        buttonColor: calculatorManager.buttonColor(button),
        isZero: button.appearance == "0"
    )
)
```

긴 파라미터는 이렇게 들여쓸 수 있습니다.

gen-com commented on week4/Calculator/Calculator/View/EachButton.swift:
```diff
@@ -0,0 +1,27 @@
+
+import SwiftUI
+
+struct EachButton: View {
```
EachButton, 네이밍 최선입니까 ?

gen-com commented on week4/Calculator/Calculator/View/PadView.swift:
```diff
@@ -0,0 +1,30 @@
+    @EnvironmentObject var calculatorManager: CalculatorManager
+    
+    var body: some View {
+        VStack {
```
Grid를 활용하면 되지 않았을까요 ?

gen-com commented on week4/Calculator/Calculator/View/ScreenView.swift:
```diff
@@ -0,0 +1,31 @@
+        HStack {
+            Spacer()
+            Text(calculatorManager.displayValue)
+                .lineLimit(1)
```
상수처리 !

gen-com commented on week4/Calculator/Calculator/Model/Calculator.swift:
```diff
@@ -0,0 +1,160 @@
+struct Calculator {
+    
+    // MARK: Alias(es)
+    
```
ViewModel의 자원을 가져와 사용하는 것 같은데, 구조를 어떻게 설계한 건가요 ?

gen-com commented on week4/Calculator/Calculator/Model/Calculator.swift:
```diff
@@ -0,0 +1,160 @@
+        }
+    }
+    
+    mutating func toggle() {
```
무엇을 토글하는 건가요 ? 좀 더 자세한 네이밍이 좋지 않을까요 ?

gen-com commented on week4/Calculator/Calculator/Model/Calculator.swift:
```diff
@@ -0,0 +1,160 @@
+
+import Foundation
+
+struct Calculator {
```
전반적으로 계산기의 모든 로직을 분석하고 설계한 것 같지는 않습니다.
읽는 것이 매끄럽지는 않으나 어느정도 이해는 할 수 있습니다.
하지만 그정도로 만족하지는 않겠죠 ?

gen-com commented on week4/Calculator/Calculator/Model/Calculator.swift:
```diff
@@ -0,0 +1,160 @@
+        }
+    }
+    
+    mutating func setOperation(_ newOperation: BinaryOperator) {
```
setOperation, equal 은 거의 유사한 로직을 가지고 있습니다. 단 한줄 차이 ! 개선할 수는 없을까요 ?

gen-com commented on week4/Calculator/Calculator/Model/Calculator.swift:
```diff
@@ -0,0 +1,160 @@
+    private(set) var isAllClear = true
+    private var calculationResult: Decimal? = 0
+    private var newValue: Decimal? = nil
+    private var operation: BinaryOperator? = .add
```
이미 Swift가 사용하는 키워드는 아래와 같이 사용 가능합니다.
``` swift
private var `operator`: BinaryOperator? = .add
```

gen-com commented on week4/Calculator/Calculator/Model/Calculator.swift:
```diff
@@ -0,0 +1,160 @@
+    typealias Digit = CalculatorManager.Digit
+    
+    // MARK: Propery(ies)
+    
```
computed를 활용할 수 있을 거 같은 느낌이 .. ?

gen-com commented on week4/Calculator/Calculator/ViewModel/CalculatorManager.swift:
```diff
@@ -0,0 +1,161 @@
+//
+//  Created by ohhyeongseok on 2022/08/16.
+//
+
```
임포트 !!!

gen-com commented on week4/Calculator/Calculator/Style/ButtonStyle.swift:
```diff
@@ -0,0 +1,42 @@
+import SwiftUI
+
+struct CalculateButtonStyle: ButtonStyle {
+    let buttonColor: (Color, Color)
```
``` swift
let buttonColor: (background: Color, foreground: Color)
```

gen-com commented on week4/Calculator/Calculator/ViewModel/CalculatorManager.swift:
```diff
@@ -0,0 +1,161 @@
+    // MARK: Property(ies)
+    
+    @Published private var calculator = Calculator()
+    let maxNumberDisplayNormalNotation: Decimal = 999999999
```
하한선은 없나요 ?

rose6649 commented on week4/Calculator/Calculator/ViewModel/CalculatorManager.swift:
```diff
@@ -7,32 +7,34 @@
+        guard let decimalTypeDisplayValue = Decimal(string: calculator.displayValue) else {
+            return "오류"
+        }
+        if decimalTypeDisplayValue > maxNumberDisplayNormalNotation {
+            return decimalTypeDisplayValue.exponentialNotation
+        }
+        return calculator.displayValue.insertComma
```
해당 문에서
string, numeric 등의 기존 타입에 extension하여 다 나눠서 코드를 구현하셨어요.
제가 생각하기로, 해당 코드를 좀 줄이고 산뜻하게 표현하고 싶어 모두 extension으로 빼신 것같아요.
그렇지만 저는 해당 구조체에서만 사용하는 프로퍼티를 그 밖의 타입까지 건들이지 않아도 된다고 생각해요.
CalculatorManager 타입 빼고는 하나도 사용하지 않는 로직, extension 프로퍼티들입니다.
사실 나중에 여러 타입, 익스텐션이 많아지게 되면 이런 기존타입 익스텐션에 있는 자투리 프로퍼티가 많아질때마다 관리가 힘들어질수도 있다고 생각하거든요.
해당 타입에서만 관련된 것들은 해당 타입안에서만 알 수 있게 하는 것도 역할에 더 맞을 것 같단 생각이 들었어요.

제 의견이니 참고만 해주시고,
네이밍이 깔끔해서 보기 좋았어요. :)

leejun6694 replied:
+
Manager 를 구현한 것은 좋은데, 이 친구가 하는 역할이 좀 많아 보입니다..!

rose6649 commented on week4/Calculator/Calculator/ViewModel/CalculatorManager.swift:
```diff
@@ -7,32 +7,34 @@
-                 [.digit(.zero), .dot, .equal]
-             ]
+    var pad: [[Button]] {
+        var buttonLayout: [[Button]] = [
+            [.allClear, .toggle, .percent, .binaryOperator(.divide)],
+            [.digit(.seven), .digit(.eight), .digit(.nine), .binaryOperator(.multiply)],
+            [.digit(.four), .digit(.five), .digit(.six), .binaryOperator(.substarct)],
+            [.digit(.one), .digit(.two), .digit(.three), .binaryOperator(.add)],
+            [.digit(.zero), .dot, .equal]
+        ]
+        if !calculator.isAllClear {
+            buttonLayout[buttonLayout.startIndex][buttonLayout[buttonLayout.startIndex].startIndex] = .clear
         }
+        return buttonLayout
```
전보다 코드가 줄어들었네요 :)

rose6649 commented on week4/Calculator/Calculator/Model/Calculator.swift:
```diff
@@ -117,6 +119,26 @@ struct Calculator {
         isAllClear = true
     }
     
+    mutating func undoWhenDragged() {
+        if newValue == nil {
+            if displayValue.count > 1 {
+                displayValue.removeLast()
+                calculationResult = Decimal(string: displayValue)
+            } else if displayValue.count == 1 {
+                displayValue = "0"
+                calculationResult = Decimal(string: displayValue)
+            }
+        } else {
+            if displayValue.count > 1 {
+                displayValue.removeLast()
+                newValue = Decimal(string: displayValue)
+            } else if displayValue.count == 1 {
+                displayValue = "0"
+                newValue = Decimal(string: displayValue)
+            }
+        }
+    }
```
코드의 중복을 줄일 수 있는 방법이 있지않을까요?

leejun6694 commented on week4/Calculator/Calculator/Model/Calculator.swift:
```diff
@@ -0,0 +1,160 @@
+        displayValue = "0"
+        newValue = nil
+        calculationResult = 0
+        operation = .add
```
isAllClear 을 여기선 안바꿔주네요?
모델이 상태값을 많이들고 있게 되는 것을 지양해보면 어떨까요?

leejun6694 commented on week4/Calculator/Calculator/Model/Calculator.swift:
```diff
@@ -0,0 +1,160 @@
+            return calculationResult - newValue
+        case .divide:
+            if newValue == 0 {
+               return nil
```
return nil 보단 에러 처리가 더 어울릴 것 같네요 : )

leejun6694 commented on week4/Calculator/Calculator/View/EachButton.swift:
```diff
@@ -0,0 +1,27 @@
+    }
+}
+
+//struct Button_Previews: PreviewProvider {
```
코드 제출 시엔 불필요한 주석을 제거해주세요.
